core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1929,4 +1961,113 @@ object Log {
   private def isLogFile(file: File): Boolean =
     file.getPath.endsWith(LogFileSuffix)
 
+  private[log] def maybeHandleOffsetOverflow[T](log: Log, segment: LogSegment, retries: Int = 1)(fn: => T): T = {
+    try fn
+    catch {
+      case e: OffsetOverflowException if (retries > 1) => {
+        info(s"Caught OffsetOverflowException $e")
+        Log.splitSegmentOnOffsetOverflow(log, segment)
+        maybeHandleOffsetOverflow(log, segment, retries - 1)(fn)
+      }
+    }
+  }
+
+  /**
+   * Split the given log segment into multiple such that there is no offset overflow in the resulting segments. The
+   * resulting segments will contain the exact same messages that are present in the input segment. On successful
+   * completion of this method, the input segment will be deleted and will be replaced by the resulting new segments.
+   * Note that if no overflow is found in the input segment, the given segment will still be replaced by a new one.
+   * See replaceSegments for recovery logic, in case the broker dies in the middle of this operation.
+   * @param log The log containing segment to split
+   * @param segment Segment to split
+   * @return List of new segments that replace the input segment
+   */
+  private[log] def splitSegmentOnOffsetOverflow(log: Log, segment: LogSegment): List[LogSegment] = {
+    val newSegments = ListBuffer[LogSegment]()
+    var position = 0
+    val sourceRecords = segment.log
+    var readBuffer = ByteBuffer.allocate(1024 * 1024)
+
+    class CopyResult(val bytesRead: Option[Int], val overflowOffset: Option[Long])
+
+    // Helper method to copy `records` into `segment`. Makes sure records being appended do not result in offset overflow.
+    def copyRecordsToSegment(records: FileRecords, segment: LogSegment, readBuffer: ByteBuffer): CopyResult = {
+      val baseOffset = segment.baseOffset
+      var bytesRead = 0
+      var maxTimestamp = Long.MinValue
+      var offsetOfMaxTimestamp = Long.MinValue
+      var maxOffset = Long.MinValue
+
+      // find all batches that are valid to be appended to the current log segment
+      val (validBatches, overflowBatches) = records.batches.asScala.span(batch => (batch.lastOffset <= baseOffset + Int.MaxValue))
```

Comment:
Should we check that ```baseOffset + Int.MaxValue``` doesn't wrap ?

Reply:
Good catch. I think we can change this to `batch.lastOffset - Int.MaxValue <= baseOffset`.